packages/exchange/src/pairs/Bridge.ts

```diff
@@ -0,0 +1,87 @@
+import Pair, { ExchangeParams, ValueTypes } from './Pair';
+import { Asset } from '@burner-wallet/types';
+import promiseRetry from 'promise-retry';
+import Web3 from "web3";
+const { toBN } = Web3.utils;
+
+
+interface BridgePairConstructor {
+    assetA: string;
+    assetABridge: string;
+    assetB: string;
+    assetBBridge: string;
+}
+
+interface BridgeExchangeParams {
+    asset: Asset;
+    assetOnOtherNetwork: Asset;
+    from: string;
+    value: string;
+    to: string;
+}
+
+export default class Bridge extends Pair {
+    protected readonly assetABridge: string;
+    protected readonly assetBBridge: string;
+
+    constructor({ assetA, assetABridge, assetB, assetBBridge }: BridgePairConstructor) {
+        super({ assetA, assetB });
+        this.assetABridge = assetABridge;
+        this.assetBBridge = assetBBridge;
+    }
+
+    exchangeAtoB({ account, value, ether }: ExchangeParams) {
+        const _value = this._getValue({ value, ether });
+        const asset = this.getExchange().getAsset(this.assetA);
+        const assetOnOtherNetwork = this.getExchange().getAsset(this.assetB);
+        return this._exchangeAndWait({
+            asset,
+            assetOnOtherNetwork,
+            from: account,
+            value: _value,
+            to: this.assetABridge
+        });
+    }
+
+    exchangeBtoA({ account, value, ether }: ExchangeParams) {
+        const _value = this._getValue({ value, ether });
+        const asset = this.getExchange().getAsset(this.assetB);
+        const assetOnOtherNetwork = this.getExchange().getAsset(this.assetA);
+        return this._exchangeAndWait({
+            asset,
+            assetOnOtherNetwork,
+            from: account,
+            value: _value,
+            to: this.assetBBridge
+        });
+    }
+
+    async _exchangeAndWait({ asset, assetOnOtherNetwork, from, value, to }: BridgeExchangeParams) {
+        const initialBalance = await assetOnOtherNetwork.getBalance(from);
+        const result = await asset.send({ from, value, to });
+        await promiseRetry(async (retry) => {
+            const updatedBalance = await assetOnOtherNetwork.getBalance(from);
+            console.log("updated balance", updatedBalance.toString())
+            if (!toBN(updatedBalance).gt(toBN(initialBalance))) {
```

Comment:
Does it mean that we will consider the transfer through the bridge finished even in a case if someone (not the bridge contract/mediator) will fill the balance with some tokens?

Does it make sense to rely on the events emitted by the bridge contract fo handle the finish of the exchange operation in more natural way?

Reply:
I updated the logic to use the bridge events to detect when the exchange operation is finished more accurately. Moved the specific logic to the `XDaiBridge` class since other bridges/mediators could have differents events.